2024-06-17/main.py

```diff
@@ -0,0 +1,201 @@
+import unittest
+import sqlite3
+import uuid
+from pprint import pprint
+from fastapi import FastAPI, APIRouter, HTTPException
+
+# Create our in-memory DB and cursor
+db = sqlite3.connect("file::memory:?cache=shared", check_same_thread=False)
+cur = db.cursor()
+
+# Create your table SQL statement
+def createDB():
+    cur.execute("CREATE TABLE users(id, name, address, email, phone)")
+
+
+# Create your table!
+createDB()
+
+# Instantiate our API/app
+app = FastAPI()
+router = APIRouter(prefix="/api")
+
+# Endpoint for adding a NEW user
+@router.post("/add/user")
+def add_user(name, address=None, email=None, phone=None) -> dict:
+    id = insertUser(name, address, email, phone)
+    return getUser(id)
+
+
+# Endpoint for getting ALL USERS
+@router.get("/get/all")
+def get_all_users(limit=10) -> list[dict]:
+    return getAllUsers(limit)
+
+
+# Endpoint for GETTING an EXISTING USER
+@router.get("/get/user")
+def get_user(id) -> dict:
+    return getUser(id)
+
+
+# Endpoint for UPDATING an EXISTING USER
+@router.patch("/update/user")
+def update_user(id, name=None, address=None, email=None, phone=None):
+    user = getUser(id)
+    if user:
+        return updateUser(
+            id,
+            user[id]["user_name"] if name is None else name,
+            user[id]["user_address"] if address is None else address,
+            user[id]["user_email"] if email is None else email,
+            user[id]["user_phone"] if phone is None else phone,
+        )
+    else:
+        raise HTTPException(status_code=404, detail="No such user.")
+
+
+# Endpoint for DELETING a USER
+@router.delete("/delete/user")
+def delete_user(id) -> str:
+    return deleteUser(id)
+
+
+# Method for CREATING a NEW USER
+def insertUser(name, address, email, phone) -> str:
+    user_id = uuid.uuid4().hex
+    params = (user_id, name, address, email, phone)
+    cur.execute(
+        "INSERT INTO users(id, name, address, email, phone) VALUES(?, ?, ?, ?, ?)",
+        params,
+    )
+    return user_id
+
+
+# Method for GETTING ALL USERS
+def getAllUsers(limit=10) -> list[dict]:
```

Comment:
Typical python style is to include the docstring as a comment under the function's signature, for example:

```python
def getAllUsers(...) -> list[dict]:
    """
    Method for GETTING ALL USERS
    """
```
If a function raises an exception, it's great to call that out in the docstring. Here's a good read: [PEP 257 – Docstring Conventions](https://peps.python.org/pep-0257/)

Reply:
I will check this out and do the needful.

Reply:
Meh 🤷‍♂️ I tried.